src/daft-parquet/src/file.rs

```diff
@@ -403,159 +411,162 @@ impl ParquetFileReader {
+                                num_values,
+                            )?;
+                            Ok(arr_iter)
+                        })
+                    });
+
+                    let arr_iters = try_join_all(arr_iter_handles)
+                        .await
+                        .context(JoinSnafu { path: uri.clone() })?
+                        .into_iter()
+                        .collect::<DaftResult<Vec<_>>>()?;
+
+                    DaftResult::Ok(arr_iters)
                 });
+                (row_range, chunk_iter_task)
+            });
+
+        let compute_runtime = get_compute_runtime();
+        let uri = self.uri.clone();
+        let parquet_task = compute_runtime.spawn(async move {
+            let mut table_tasks = Vec::with_capacity(chunk_iter_handles.len());
+            for ((row_range, chunk_iter_handle), output_sender) in chunk_iter_handles.zip(senders) {
+                // We want to ensure that the channel capacity can hold one morsel worth of data for better deserialization performance.
+                let channel_size = {
+                    let chunks_per_morsel = max(PARQUET_MORSEL_SIZE / chunk_size, 1);
+                    let max_num_chunks = max(row_range.num_rows / chunk_size, 1);
+                    min(chunks_per_morsel, max_num_chunks)
+                };
 
-        let _ = try_join_all(table_iter_handles)
-            .await
-            .context(JoinSnafu { path: self.uri })?
-            .into_iter()
-            .collect::<DaftResult<Vec<_>>>()?;
+                let chunk_iter = chunk_iter_handle
+                    .await
+                    .context(JoinSnafu { path: uri.clone() })??;
+
+                let permit = semaphore.clone().acquire_owned().await.unwrap();
+                let table_task = spawn_column_iters_to_table_task(
+                    chunk_iter,
+                    row_range,
+                    daft_schema.clone(),
+                    uri.clone(),
+                    predicate.clone(),
+                    original_columns.clone(),
+                    original_num_rows,
+                    delete_rows.clone(),
+                    output_sender,
+                    permit,
+                    channel_size,
+                );
+                table_tasks.push(table_task);
+            }
 
-        let combined_stream =
-            futures::stream::iter(receivers.into_iter().map(futures::stream::iter));
-        match maintain_order {
-            true => Ok(Box::pin(combined_stream.flatten())),
-            false => Ok(Box::pin(combined_stream.flatten_unordered(None))),
-        }
+            try_join_all(table_tasks)
+                .await?
+                .into_iter()
+                .collect::<DaftResult<()>>()?;
+            DaftResult::Ok(())
+        });
+
+        let stream_of_streams =
+            futures::stream::iter(receivers.into_iter().map(ReceiverStream::new));
+        let flattened = match maintain_order {
```

Comment:
we ideally don't want to have multiple boxes for `dyn Iterator`. The code then cannot be inlined.

I suggest making `combined_stream` a generic nested function and then do

```            
true => combine_stream(stream_of_streams.flatten()).boxed(),
```

so that we only have 1 boxed iterator